libs/langchain/langchain/llms/bedrock.py

```diff
@@ -26,7 +27,21 @@ def prepare_input(
         cls, provider: str, prompt: str, model_kwargs: Dict[str, Any]
     ) -> Dict[str, Any]:
         input_body = {**model_kwargs}
-        if provider == "anthropic" or provider == "ai21":
+        if provider == "anthropic":
```

Comment:
The logic in this branch is not correct. For example (abstracting it into a function):

```
def format_prompt(prompt):
      if prompt.startswith("Human:") and prompt.endswith("Assistant:"):
          reformatted_prompt = prompt  # Already wrapped
      else:
          # Guard against common errors in specifying wrong number of newlines.
          corrected_prompt, n_subs = re.subn(r"^\n*Human:", "Human:", prompt)
          if n_subs == 1:
              reformatted_prompt = corrected_prompt

          # As a last resort, wrap the prompt ourselves to emulate instruct-style.
          else:
              reformatted_prompt = f"{'Human:'} {prompt}{'Assistant:'}"
      return reformatted_prompt

format_prompt("Hello")
> 'Human: Hello.Assistant:'
```

Reply:
Here's some prompt-reformatting code I wrote for Bedrock that is being implemented into their error handling.

```HUMAN_PROMPT = "\n\nHuman:"
ASSISTANT_PROMPT = "\n\nAssistant:"
ALTERNATION_ERROR = "Error: Prompt must alternate between '\n\nHuman:' and '\n\nAssistant:'."


def add_newlines_before_ha(inputText):
    new_text = inputText
    for word in ["Human:", "Assistant:"]:
        new_text = new_text.replace(word, "\n\n" + word)
        for i in range(2):
            new_text = new_text.replace("\n\n\n" + word, "\n\n" + word)
    return new_text
            

def human_assistant_format(inputText):
    if inputText.count("Human:") == 0 or (inputText.find("Human:") > inputText.find("Assistant:") and "Assistant:" in inputText):
        inputText = HUMAN_PROMPT + ' ' + inputText  # SILENT CORRECTION
    if inputText.count("Assistant:") == 0:
        inputText = inputText + ASSISTANT_PROMPT  # SILENT CORRECTION
    if inputText[:len("Human:")] == "Human:":
        inputText = "\n\n" + inputText
    inputText = add_newlines_before_ha(inputText)
    count = 0
    # track alternation
    for i in range(len(inputText)):
        if inputText[i:i + len(HUMAN_PROMPT)] == HUMAN_PROMPT:
            if count % 2 == 0:
                count += 1
            else:
                return ALTERNATION_ERROR  # Error
        if inputText[i:i + len(ASSISTANT_PROMPT)] == ASSISTANT_PROMPT:
            if count % 2 == 1:
                count += 1
            else:
                return ALTERNATION_ERROR  # ERROR

    if count % 2 == 1:  # Only saw Human, no Assistant
        inputText = inputText + ASSISTANT_PROMPT  # SILENT CORRECTION

    return inputText
```

And here are some test cases.

test_cases = {
"""Hey""": """

Human: Hey

Assistant:""",






"""

Human: Hello

Assistant:""": """

Human: Hello

Assistant:""",



"""Human: Hello

Assistant:""": """

Human: Hello

Assistant:""",



"""
Human: Hello

Assistant:""": """

Human: Hello

Assistant:""",



"""

Human: Human: Hello

Assistant:""": "Error: Prompt must alternate between '\n\nHuman:' and '\n\nAssistant:'.",



"""Human: Hello

Assistant: Hello

Human: Hello

Assistant:""": """

Human: Hello

Assistant: Hello

Human: Hello

Assistant:""",





"""

Human: Hello

Assistant: Hello

Human: Hello

Assistant:""": """

Human: Hello

Assistant: Hello

Human: Hello

Assistant:""",



"""

Human: Hello

Assistant: Hello

Human: Hello

Assistant: Hello

Assistant: Hello""": ALTERNATION_ERROR,




"""

Human: Hi.

Assistant: Hi.

Human: Hi.

Human: Hi.

Assistant:""": ALTERNATION_ERROR,






"""
Human: Hello""": """

Human: Hello

Assistant:""",





"""

Human: Hello
Hello

Assistant""": """

Human: Hello
Hello

Assistant

Assistant:""",





"""Hello

Assistant:""": """

Human: Hello

Assistant:""",






"""Hello

Human: Hello

""": """Hello

Human: Hello



Assistant:""",







"""

Human: Assistant: Hello""": """

Human: 

Assistant: Hello""",




"""

Human: Human

Assistant: Assistant

Human: Assistant

Assistant: Human""": """

Human: Human

Assistant: Assistant

Human: Assistant

Assistant: Human""",






"""
Assistant: Hello there, your name is:

Human.

Human: Hello there, your name is: 

Assistant.""": """

Human: 

Assistant: Hello there, your name is:

Human.

Human: Hello there, your name is: 

Assistant.

Assistant:""",





"""

Human: Human: Hi

Assistant: Hi""": ALTERNATION_ERROR,




"""Human: Hi

Human: Hi""": ALTERNATION_ERROR,





"""

Assistant: Hi

Human: Hi""": """

Human: 

Assistant: Hi

Human: Hi

Assistant:""",







"""

Human: Hi

Assistant: Yo

Human: Hey

Assistant: Sup

Human: Hi

Assistant: Hi
Human: Hi
Assistant:""": """

Human: Hi

Assistant: Yo

Human: Hey

Assistant: Sup

Human: Hi

Assistant: Hi

Human: Hi

Assistant:""",




"""

Hello.

Human: Hello.

Assistant:""": """

Hello.

Human: Hello.

Assistant:"""
}

for p in test_cases:
    assert(human_assistant_format(p) == test_cases[p])